format-common/src/main/java/co/cask/hydrator/format/plugin/AbstractFileSink.java

```diff
@@ -0,0 +1,127 @@
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    config.validate();
+  }
+
+  @Override
+  public final void prepareRun(BatchSinkContext context) {
+    config.validate();
+
+    // set format specific properties.
+    outputFormatter = config.getFormat().getFileOutputFormatter(config.getProperties().getProperties(),
+                                                                config.getSchema());
+    if (outputFormatter == null) {
+      // should never happen, as validation should enforce the allowed formats
+      throw new IllegalArgumentException(String.format("Format '%s' cannot be used to write data.",
+                                                       config.getFormat()));
+    }
+
+    // record field level lineage information
+    // needs to happen before context.addOutput(), otherwise an external dataset without schema will be created.
+    Schema schema = config.getSchema();
+    if (schema == null) {
+      schema = context.getInputSchema();
+    }
+    LineageRecorder lineageRecorder = new LineageRecorder(context, config.getReferenceName());
+    lineageRecorder.createExternalDataset(schema);
+    if (schema != null && schema.getFields() != null && !schema.getFields().isEmpty()) {
+      recordLineage(lineageRecorder,
+                    schema.getFields().stream().map(Schema.Field::getName).collect(Collectors.toList()));
+    }
+
+    Map<String, String> outputProperties = new HashMap<>(outputFormatter.getFormatConfig());
+    outputProperties.putAll(getFileSystemProperties(context));
+    outputProperties.put(FileOutputFormat.OUTDIR, getOutputDir(context.getLogicalStartTime()));
+
+    context.addOutput(Output.of(config.getReferenceName(),
+                                new SinkOutputFormatProvider(outputFormatter.getFormatClassName(), outputProperties)));
+  }
+
+  @Override
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    super.initialize(context);
+    outputFormatter = config.getFormat().getFileOutputFormatter(config.getProperties().getProperties(),
+                                                                config.getSchema());
+  }
+
+  @Override
+  public void transform(StructuredRecord input, Emitter<KeyValue<Object, Object>> emitter) throws Exception {
+    emitter.emit(outputFormatter.transform(input));
+  }
+
+  /**
+   * Override this to provide any additional Configuration properties that are required by the FileSystem.
+   * For example, if the FileSystem requires setting properties for credentials, those should be returned by
+   * this method.
+   */
+  protected Map<String, String> getFileSystemProperties(BatchSinkContext context) {
+    return Collections.emptyMap();
+  }
+
+  /**
+   * Override this to specify a custom field level operation name and description.
+   */
+  protected void recordLineage(LineageRecorder lineageRecorder, List<String> outputFields) {
+    lineageRecorder.recordRead("Read", "Read from files.", outputFields);
```

Comment:
here can we say "Read from %s files." where %s comes from from the config.format ?

Reply:
done